crates/compiler/mono/src/ir.rs

```diff
@@ -3191,19 +3200,25 @@ fn specialize_external<'a>(
                                 size2.cmp(&size1)
                             });
 
+                            let ordered_field_layouts = Vec::from_iter_in(
+                                combined.iter().map(|(_, layout)| **layout),
+                                env.arena,
+                            );
+                            let ordered_field_layouts = ordered_field_layouts.into_bump_slice();
+
                             debug_assert_eq!(
                                 captured.len(),
-                                field_layouts.len(),
+                                ordered_field_layouts.len(),
                                 "{:?} captures {:?} but has layout {:?}",
                                 lambda_name,
                                 &captured,
-                                &field_layouts
+                                &ordered_field_layouts
                             );
 
                             for (index, (symbol, layout)) in combined.iter().enumerate() {
                                 let expr = Expr::StructAtIndex {
                                     index: index as _,
-                                    field_layouts,
+                                    field_layouts: ordered_field_layouts,
                                     structure: Symbol::ARG_CLOSURE,
                                 };
```

Comment:
As @brian-carroll found, we just need to do a good sort of the field layouts here.

Reply:
and you want to do this in a separate PR?

Reply:
this updated it so that we use the sorted field layouts. The TODOs we can address in a separate PR.